File: 2nd-project/sudoku.pl
% ============= Sudoku Solver in Prolog ==============
% See: https://en.wikipedia.org/wiki/Sudoku
% ====================================================
% @ISEL, 2022-2023 Summer Semester
% BSc in Computer Science and Engineering
% Artificial Intelligence
% ====================================================
% Authors:
%   - 48280 André Jesus
%   - 48287 Nyckollas Brandão
% Professor:
%   - Nuno Leite
% ====================================================
% To run the solver, use this query: 
% ?- sudoku.
% ====================================================

:- consult('utils.pl').
:- consult('goal.pl').
:- consult('iterative_deepening/iddfs_sudoku.pl').
:- consult('bestfirst/bestfirst_sudoku.pl').
:- use_module(library(clpfd)).

% Runs the sudoku solver
sudoku :-
    writeln('Welcome to the Sudoku Solver in Prolog!'),
    puzzle_difficulty(Difficulty),
    puzzle(Difficulty, Puzzle),
    writeln('Puzzle:'),
    print_puzzle(Puzzle),
    puzzle_algorithm(Algorithm),
    writeln('Solving...'),
    solve(Puzzle, Algorithm, Solution),
    writeln('Solution:'),
    print_puzzle(Solution).

% Asks the user for the difficulty of the puzzle
puzzle_difficulty(Difficulty) :-
    writeln('Difficulties:'),
    writeln('0 - Easy'),
    writeln('1 - Medium'),
    writeln('2 - Hard'),
    writeln('3 - Very Hard'),
    write('Choose the difficulty of the puzzle [0-3]: '),
    read(Difficulty),
    between(0, 3, Difficulty).

% Asks the user for the algorithm to solve the puzzle
puzzle_algorithm(Algorithm) :-
    writeln('Algorithms:'),
    writeln('0 - Iterative Deepening Depth-First Search (IDDFS)'),
    writeln('1 - Best-First Search (A*)'),
    write('Choose the algorithm to solve the puzzle [0-1]: '),
    read(Algorithm),
    between(0, 1, Algorithm).

% Returns puzzles with different difficulty levels
% puzzle(0, Puzzle) - Easy
% puzzle(1, Puzzle) - Medium
% puzzle(2, Puzzle) - Hard
% puzzle(3, Puzzle) - Very Hard
puzzle(0, [ [_,_,5,2,6,9,7,8,1],
            [6,8,2,5,7,1,4,9,3],
            [1,_,7,8,3,4,5,6,2],
            [8,2,6,1,_,5,3,_,7],
            [3,7,4,6,8,2,9,1,5],
            [9,5,1,7,4,3,6,2,8],
            [5,1,9,_,2,6,8,7,4],
            [2,4,8,9,5,7,1,3,6],
            [7,6,3,4,1,8,2,5,_] ]).

puzzle(1, [ [5,_,_,6,_,1,_,_,_],
            [_,3,_,_,7,5,_,4,9],
            [_,_,_,9,4,8,_,_,_],
            [1,5,7,_,_,_,_,_,_],
            [_,9,6,_,_,_,2,_,8],
            [2,_,_,1,6,9,_,5,_],
            [4,1,_,3,_,7,_,6,_],
            [_,2,_,5,1,_,3,7,_],
            [7,_,3,4,_,_,1,8,_] ]).

puzzle(2, [ [_,_,_,6,_,_,4,_,_],
            [7,_,_,_,_,3,6,_,_],
            [_,_,_,_,9,1,_,8,_],
            [_,_,_,_,_,_,_,_,_],
            [_,5,_,1,8,_,_,_,3],
            [_,_,_,3,_,6,_,4,5],
            [_,4,_,2,_,_,_,6,_],
            [9,_,3,_,_,_,_,_,_],
            [_,2,_,_,_,_,1,_,_] ]).

puzzle(3, [ [_,_,_,_,_,_,_,_,_],
            [_,_,_,_,_,3,_,8,5],
            [_,_,1,_,2,_,_,_,_],
            [_,_,_,5,_,7,_,_,_],
            [_,_,4,_,_,_,1,_,_],
            [_,9,_,_,_,_,_,_,_],
            [5,_,_,_,_,_,_,7,3],
            [_,_,2,_,1,_,_,_,_],
            [_,_,_,_,4,_,_,_,9] ]). 

% ==============================================
% =============== Print Puzzle =================
% ==============================================

% Prints the puzzle
print_puzzle([]).
print_puzzle([H|T]) :-
    print_puzzle_r([H|T], 1).

% Prints the puzzle recursively
print_puzzle_r([], _).
print_puzzle_r([H|T], N) :-
    write(' '),
    print_row(H, 1),
    ((N == 3; N == 6) 
        -> writeln('\n-------+-------+-------')
        ;  writeln('')
    ),
    N1 is N + 1,
    print_puzzle_r(T, N1).

% Prints a row of the puzzle
print_row([], _).
print_row([H|T], N) :-
    (H = '_' -> write(' '); write(H)),
    ((N == 3; N == 6) 
        -> write(' | ')
        ;  write(' ')
    ),
    N1 is N + 1,
    print_row(T, N1).

% ==============================================
% =============== Solve Puzzle =================
% ==============================================

% Solves the puzzle using iterative deepening depth-first search (IDDFS)
solve(Puzzle, 0, Solution) :-
    depth_first_iterative_deepening(Puzzle, [Solution | _], 1000). % TODO: sus depth (use number of missing cells?)

% Solves the puzzle using best-first search (A*)
solve(Puzzle, 1, Solution) :-
    bestfirst(Puzzle, [Solution | _]).

File: 2nd-project/matlab/sudoku.m
% ============= Sudoku Solver in MATLAB ==============
% See: https://en.wikipedia.org/wiki/Sudoku
% ====================================================
% @ISEL, 2022-2023 Summer Semester
% BSc in Computer Science and Engineering
% Artificial Intelligence
% ====================================================
% Authors:
%   - 48280 André Jesus
%   - 48287 Nyckollas Brandão
% Professor:
%   - Nuno Leite
% ====================================================
% To run the application, run the "sudoku" function
% ====================================================

source('./simulated_annealing/sa_sudoku.m');
source('./genetic_algorithm/ga_sudoku.m');

% Function: sudoku
% ----------------------------
% Entry point of the application
function sudoku()
    disp('Welcome to the Sudoku Solver in MATLAB!');

    disp('Difficulties:')
    disp('0 - Easy');
    disp('1 - Medium');
    disp('2 - Hard');
    disp('3 - Expert');
    disp('4 - Nightmare');
    disp('5 - Impossible')
    difficulty = input('Choose the difficulty of the puzzle [0-5]: ');
    while ~(difficulty >= 0 && difficulty <= 5)
        difficulty = input('Invalid input! Please choose a difficulty between 0 and 5: ');
    end

    puzzle = puzzle(difficulty);

    disp('Puzzle:');
    print_puzzle(puzzle);

    disp('Algorithms:')
    disp('0 - Simulated Annealing');
    disp('1 - Genetic Algorithm');
    algorithm = input('Choose the algorithm to solve the puzzle [0-1]: ');
    while ~(algorithm >= 0 && algorithm <= 1)
        algorithm = input('Invalid input! Please choose an algorithm between 0 and 1: ');
    end

    disp('Solving...');
    solution = solve(puzzle, algorithm);

    disp('Solution:');
    print_puzzle(solution);
end

% Function: print_puzzle
% ----------------------------
% Prints the puzzle in a human-readable format
%
% @param puzzle: the puzzle to print
function print_puzzle(puzzle)
    for i = 1:9
        fprintf('%s', ' ');
        for j = 1:9
            if isnan(puzzle(i, j))
                fprintf('%s', ' ');
            else
                fprintf('%d', puzzle(i, j));
            end
            if mod(j, 3) == 0 && j ~= 9
                fprintf('%s', ' | ');
            else
                fprintf('%s', ' ');
            end
        end
        disp(' ');
        if mod(i, 3) == 0 && i ~= 9
            disp('-------+-------+-------');
        end
    end
    disp(' ');
end

% Function: solve
% ----------------------------
% Solves the puzzle using the specified algorithm
%
% @param puzzle: the puzzle to solve
% @param algorithm: the algorithm to use
function solution = solve(puzzle, algorithm)
    % Create struct data that will contain problem data
    optimum = 0;
    data = struct('puzzle', puzzle, 'optimum', optimum);
    sense = 'minimize'; % Minimization problem

    if algorithm == 0
        t_max = 5000;      % Initial temperature (can be adjusted)
        t_min = 0.001;     % Final temperature (can be adjusted)
        r = 0.005;         % Cooling rate (can be adjusted)
        k = 8;             % Number of iterations per temperature (can be adjusted)

        results = sa(t_max, t_min, r, k, data, @get_initial_solution, @get_random_neighbour, @evaluate, ...
            @is_optimum, sense);

        res = [[results(:).t]; [results(:).num_evaluations]; [results(:).cost]];
        fprintf('Final Temp\t\tNum Evaluations\t\tBest Cost\n');
        fprintf('%.7f\t\t%d\t\t%d\n', res);

        solution = results(end).s;
    else
        t_max = [1000];      % Max number of iterations
        pop_size = [100];    % Population size
        cross_prob = [0.8];  % Cross  probability
        mut_prob = [0.25];   % Mutation probability
        num_of_tests = length(t_max);

        % Run Tests
        for f = 1 : num_of_tests
            fprintf('\ntmax=%d', t_max(f));
            fprintf('\npopSize=%d', pop_size(f));
            fprintf('\ncrossProb=%.1f', cross_prob(f));
            fprintf('\nmutProb=%.1f\n', mut_prob(f));

            fprintf('\nNumEvaluations\t\tCost\n');
            fprintf('============================\n');
            results = ga(data, t_max(f), pop_size(f), cross_prob(f), mut_prob(f), @select, @crossover, @mutate, ...
                @get_initial_solution, @evaluate, @is_optimum, sense);

            res = [[results.num_evaluations]; [results.cost]];
            fprintf('%d\t\t%d\n', res);

            solution = results.s;
        end
    end
end

